source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -76,73 +124,184 @@ FilterHeadersStatus Filter::encodeHeaders(ResponseHeaderMap& headers, bool end_o
   auto* headers_req = req.mutable_response_headers();
   MutationUtils::buildHttpHeaders(headers, *headers_req->mutable_headers());
   headers_req->set_end_of_stream(end_of_stream);
-  response_state_ = FilterState::HEADERS;
+  response_state_ = FilterState::Headers;
   ENVOY_LOG(debug, "Sending response_headers message");
   stream_->send(std::move(req), false);
   stats_.stream_msgs_sent_.inc();
+  ENVOY_LOG(trace, "encodeHeaders: StopAllIterationAndWatermark");
   return FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
-void Filter::onReceiveMessage(
-    std::unique_ptr<envoy::service::ext_proc::v3alpha::ProcessingResponse>&& r) {
+FilterDataStatus Filter::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(trace, "encodeData({}): end_of_stream = {}", data.length(), end_stream);
+  if (processing_complete_) {
+    ENVOY_LOG(trace, "encodeData: Continue (complete)");
+    return FilterDataStatus::Continue;
+  }
+
+  switch (processing_mode_.response_body_mode()) {
+  case ProcessingMode::BUFFERED:
+    if (end_stream) {
+      ENVOY_BUG(response_state_ == FilterState::Idle, "Invalid filter state on response path");
+      if (encoder_callbacks_->encodingBuffer() == nullptr) {
+        // Whole body in one chunk
+        response_state_ = FilterState::ChunkedBody;
+        response_body_chunk_ = &data;
+        sendBodyChunk(false, data, true);
+        return FilterDataStatus::StopIterationNoBuffer;
+      } else {
+        // Multiple chunks that were buffered
+        response_state_ = FilterState::BufferedBody;
+        sendBodyChunk(false, *encoder_callbacks_->encodingBuffer(), true);
+        return FilterDataStatus::StopIterationAndBuffer;
```

Comment:
Can we simplify this case branching by using `addDecodedDate()` here instead?

Reply:
Yes -- I can see how that helps. I used it and I was able to streamline the logic a lot as you will see.